### PR TITLE
Obstacle avoidance bug fix in guided mode

### PR DESCRIPTION
**Pull request by sharvashish**
This PR is trying to solve #7253 .
added code to adjust z velocity using avoidance
also, added velocity adjust code using avoidance, when desired velocity is same as that of current desired

### CODE REVIEW

**Comment by rmackay9** [comment]
Hi there, thanks for the fix.  I see there's some discussion with @OXINARF on the issue.

I think we can just remove these few lines:
     // exit immediately if already equal
     if (curr_vel_des == vel_des) {

.. and then the other change you've done to add a call to "get_avoidance_adjusted_climbrate".

**Comment by sharvashish** [comment]
@rmackay9 Thank you for acknowledging this.

> I think we can just remove these few lines:
> // exit immediately if already equal
> if (curr_vel_des == vel_des) {
> return;
> }

So what do you suggest about this? Should I create a new commit with this change of logic in code?

File: ArduCopter/control_guided.cpp [hunk 1]
```diff
@@ -663,6 +663,14 @@ void Copter::guided_set_desired_velocity_with_accel_and_fence_limits(const Vecto
 
     // exit immediately if already equal
     if (curr_vel_des == vel_des) {
```
Comment (OXINARF) [inline ArduCopter/control_guided.cpp]: My proposal for this was:

```
if (curr_vel_des != vel_des) {

    // get change in desired velocity
    Vector3f vel_delta = vel_des - curr_vel_des;

    [...]
}

#if AC_AVOID_ENABLED
[...]
```

I see no reason to run the delta math calculations when delta will be zero.

File: ArduCopter/control_guided.cpp [hunk 2]
```diff
@@ -686,6 +694,8 @@ void Copter::guided_set_desired_velocity_with_accel_and_fence_limits(const Vecto
 #if AC_AVOID_ENABLED
     // limit the velocity to prevent fence violations
     avoid.adjust_velocity(pos_control->get_pos_xy_kP(), pos_control->get_accel_xy(), curr_vel_des);
+    // get avoidance adjusted climb rate
+    curr_vel_des.z = get_avoidance_adjusted_climbrate(curr_vel_des.z);
```
Comment (OXINARF) [inline ArduCopter/control_guided.cpp]: It's up to @rmackay9, but I would directly do ` avoid.adjust_velocity_z(pos_control->get_pos_z_kP(), pos_control->get_accel_z(), curr_vel_des.z);`.

The `get_avoidance_adjusted_climbrate` is useful for other flight modes where we don't need to put macro checks, here we already have the macro check.

Comment (sharvashish) [reply ArduCopter/control_guided.cpp]: Yes @OXINARF , you are absolutely right about this. My bad, I should have seen this.

File: ArduCopter/control_guided.cpp (hunk 1 above)
Comment (sharvashish) [reply ArduCopter/control_guided.cpp]: Yes @OXINARF , it definitely makes sense. @rmackay9 eagerly awaiting your response.

**Comment by sharvashish** [comment]
Hello @OXINARF @rmackay9 ,

Any update on this ?
Should I edit this PR?

**Comment by OXINARF** [comment]
@sharvashish Go ahead and do the changes I proposed.

**Comment by rmackay9** [comment]
Hi there.  Sorry for the conflicting messages from @OXINARF and I.  there are, as per usual, multiple acceptable ways to do it.

I personally prefer simpler code even if it means doing a couple of extra calculations so I've taken your change but removed the "if (curr_vel_des != vel_des) {" part.

**Comment by rmackay9** [comment]
By the way, I also adjusted the commit message to match our regular format.  i.e. added "Copter:" at the front.

**Comment by sharvashish** [comment]
Hi @rmackay9 , Thanks for including this bug fix.
Can you please direct me to commit rules of this codebase so that I don't make the same mistake again.

**Comment by khancyr** [comment]
@sharvashish http://ardupilot.org/dev/docs/contributing.html ;-P

**Comment by sharvashish** [comment]
@khancyr Thank You for sharing the link. :)

**Pull request closed without merging**